spring-cloud-gcp-samples/spring-cloud-gcp-storage-resource-sample/src/main/java/com/example/WebController.java

```diff
@@ -35,19 +39,48 @@
 @RestController
 public class WebController {
 
+  @Value("${gcs-resource-test-bucket}")
+  private String bucketName;
+
   @Value("gs://${gcs-resource-test-bucket}/my-file.txt")
   private Resource gcsFile;
 
-  @GetMapping("/")
-  public String readGcsFile() throws IOException {
+  private Storage storage;
+
+  private WebController(Storage storage) {
+    this.storage = storage;
+  }
+
+  @GetMapping(value = "/")
+  public String readGcsFile(@RequestParam("filename") Optional<String> filename)
+      throws IOException {
+    if (filename.isPresent()) {
+      GoogleStorageResource resource = fetchResource(filename);
+      return StreamUtils.copyToString(resource.getInputStream(), Charset.defaultCharset()) + "\n";
+    }
     return StreamUtils.copyToString(this.gcsFile.getInputStream(), Charset.defaultCharset()) + "\n";
   }
 
-  @PostMapping("/")
-  public String writeGcs(@RequestBody String data) throws IOException {
-    try (OutputStream os = ((WritableResource) this.gcsFile).getOutputStream()) {
+  @PostMapping(value = "/")
+  public String writeGcs(
+      @RequestBody String data, @RequestParam("filename") Optional<String> filename)
+      throws IOException {
+    if (filename.isPresent()) {
+      GoogleStorageResource resource = fetchResource(filename);
+      return updateResource(resource, data);
+    }
+    return updateResource(this.gcsFile, data);
+  }
+
+  private String updateResource(Resource resource, String data) throws IOException {
+    try (OutputStream os = ((WritableResource) resource).getOutputStream()) {
       os.write(data.getBytes());
     }
     return "file was updated\n";
   }
+
+  private GoogleStorageResource fetchResource(Optional<String> filename) {
+    return new GoogleStorageResource(
+        this.storage, String.format("gs://%s/%s", this.bucketName, filename.get()));
```

Comment:
This is a private method, so you know that you've only invoked this `get()` in a different method -- but this method's contract doesn't say it's invalid to pass in an Optional.empty(), it says the opposite: "You can pass any optional here."

Since this would throw an NPE if you passed in Optional.empty(), please change this to either check isPresent() again, or accept `String` as the parameter, and invoke get in the same method in which you checked isPresent().